pinot-core/src/main/java/org/apache/pinot/core/data/table/OrderedIndexedTableResizer.java

```diff
@@ -0,0 +1,260 @@
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.PriorityQueue;
+import java.util.Set;
+import org.apache.commons.collections.comparators.ComparableComparator;
+import org.apache.pinot.common.request.AggregationInfo;
+import org.apache.pinot.common.request.SelectionSort;
+import org.apache.pinot.common.utils.DataSchema;
+import org.apache.pinot.core.query.aggregation.function.AggregationFunction;
+import org.apache.pinot.core.query.aggregation.function.AggregationFunctionUtils;
+
+
+/**
+ * Helper class for trimming and sorting records in the IndexedTable, based on the order by information
+ */
+class OrderedIndexedTableResizer extends IndexedTableResizer {
+
+  private List<OrderByInfo> _orderByInfos;
+  private Comparator<IntermediateRecord> _intermediateRecordComparator;
+  private int _numOrderBy;
+
+  OrderedIndexedTableResizer(DataSchema dataSchema, List<AggregationInfo> aggregationInfos,
+      List<SelectionSort> orderBy) {
+
+    int numAggregations = aggregationInfos.size();
+    int numKeyColumns = dataSchema.size() - numAggregations;
+
+    Map<String, Integer> keyIndexMap = new HashMap<>();
+    for (int i = 0; i < numKeyColumns; i++) {
+      keyIndexMap.put(dataSchema.getColumnName(i), i);
+    }
+
+    Map<String, Integer> aggregationColumnToIndex = new HashMap<>();
+    Map<String, AggregationInfo> aggregationColumnToInfo = new HashMap<>(aggregationInfos.size());
+    for (int i = 0; i < numAggregations; i++) {
+      AggregationInfo aggregationInfo = aggregationInfos.get(i);
+      String aggregationColumn = AggregationFunctionUtils.getAggregationColumnName(aggregationInfo);
+      aggregationColumnToIndex.put(aggregationColumn, i);
+      aggregationColumnToInfo.put(aggregationColumn, aggregationInfo);
+    }
+
+    _numOrderBy = orderBy.size();
+    _orderByInfos = new ArrayList<>(_numOrderBy);
+    Comparator[] comparators = new Comparator[_numOrderBy];
+
+    for (int i = 0; i < _numOrderBy; i++) {
+      SelectionSort selectionSort = orderBy.get(i);
+      String column = selectionSort.getColumn();
+
+      if (keyIndexMap.containsKey(column)) {
+        int index = keyIndexMap.get(column);
+        _orderByInfos.add(createOrderByInfoForKeyColumn(index));
+      } else if (aggregationColumnToIndex.containsKey(column)) {
+        int index = aggregationColumnToIndex.get(column);
+        _orderByInfos.add(createOrderByInfoForAggregationColumn(index, aggregationColumnToInfo.get(column)));
+      } else {
+        throw new UnsupportedOperationException("Could not find column " + column + " in data schema");
+      }
+
+      comparators[i] = ComparableComparator.getInstance();
+      if (!selectionSort.isIsAsc()) {
+        comparators[i] = comparators[i].reversed();
+      }
+    }
+
+    _intermediateRecordComparator = (o1, o2) -> {
+
+      for (int i = 0; i < _numOrderBy; i++) {
+        int result = comparators[i].compare(o1.getValues()[i], o2.getValues()[i]);
```

Comment:
Cache o1.getValues() and o2.getValues() (or you can directly access the array if you make the class inner class)